### ISSUE TITLE
Populate hostvars with the host_vars variables of every node

### ISSUE THREAD

**Issue by Bengrunt** [ansible/ansible#11503]
Hey there,

More like a feature request than an issue here.
It could be nice to have the variables located in the host_vars directory copied in the `hostvars` variable as well so that a single node could access other nodes specific variables.

For instance, when deploying a service cluster with a playbook you might need the cluster frontend to be able to generate its config files with other node specific information that is not retrieved by the ansible setup (i.e. cluster technology specific information) but located in the host_vars. Acutally, I'm writing a playbook to deploy a postgresql cluster with pgpool as the frontend in that fashion.

In hostvars, there is already an host based layout:

```
"node1" : {
    setup information from node1...
},
"node2": {
    setup information from node2...
},
...
"group_vars": {
    group variables for current node
},
...
"some_host_vars_variable_for_current_node": "some value",
"some_host_vars_variable_for_current_node": "some value",
"some_host_vars_variable_for_current_node": "some value"
```

For each node involved in the playbook call, we can access the facts retrieved during setup. I'd like to be able to access group_vars and host_vars in the same way but can only access those of the currently running node.

I know I could just write all the information in the group_vars in a very static way but I didn't find it scalable nor elegant. Maybe there is something to do that already but I didn't find it :/

Thanks a lot for your help.

**Comment by bcoca** [comment]
actually you can already access hostvars from other hosts, hostvars['otherhost'].varname, there are no group_vars or host_vars keys as they are flattened to the 'varname' when inventory is read.

Other variables that are populated into hostvars come from running fact gathering on those nodes, those won't be available unless you gathered facts on them or are using fact caching.

**Comment by Bengrunt** [comment]
Well, If I do something like:

```
...
host_vars:
  - node1: var1=value1 # All that is in a file obviously.
```

If I look into hostvars when running on node2, there is no such key as hostvars['node1'].var1.
Should there be one?

I'm running Ansible v1.9.1 on CentOS7 and my targets are Ubuntu 14.04LTS hosts.

**Comment by Bengrunt** [comment]
I could access a value in hostvars.var1 but that would only give me the value set in the host_vars file of node2 if it exists and the value is set. So, nothing from the other hosts sadly.

**Comment by bcoca** [comment]
I think you are misunderstanding how it works, that is your own construct, host_vars in ansible is a directory that contains per host files which then get assigned to that host. 

your construct above will create a host_vars variable with a list, so you won't be able to access by node name, also the var1=value1 will not be parsed, they are the value of the dictionary key 'node1'

**Comment by bcoca** [comment]
so closing this as it is not a feature request but a misunderstanding

**Comment by bcoca** [comment]
# List Information

Hi!

Thanks very much for your interest in Ansible.  It sincerely means a lot to us. 

This appears to be a user question, and we'd like to direct these kinds of things to either the mailing list or the IRC channel.
- IRC: #ansible on irc.freenode.net  
- mailing list: https://groups.google.com/forum/#!forum/ansible-project

If you can stop by there, we'd appreciate it.  This allows us to keep the issue tracker for bugs, pull requests, RFEs and the like.

Thank you once again and we look forward to seeing you on the list or IRC.  Thanks!

**Comment by Bengrunt** [comment]
I'm sorry, I think I chose a really bad way to represent my host_vars directory.
I didn't mean that I declare an host_vars variable in my playbook, just that I have a variable file for a specific host in the host_vars directory of the playbook.

So first, I set a variable `test` in the yml file .../host_vars/node1.yml:

```
test: "some value there"
```

Then, I've got some task that is run by node2 only, with a `when` statement for instance. In this task, I need to template a file with the value contained in the variable `test` that is only defined for node1.
I would have thought that it would indeed be accessible through the hostvars variable, at `hostvars['node1'].test`, but that's not the case.

**Comment by bcoca** [comment]
I just tested and that works fine for me

host_vars/localhost2: temp: test

```
- hosts: localhost
  gather_facts: false
  tasks:
    - debug: msg={{hostvars['localhost2']["temp"]}}
```

i ran and got:

```
ok: [localhost] => {
    "msg": "test", 
    "changed": false
}
```

**Comment by Bengrunt** [comment]
Okay, so I tried your example and it worked as expected.
So I digged a little bit more to understand what I could have missed and it appears that there is a big difference between `hostvars` and `hostvars['some_host']`.

The first contains information about all nodes but does not include all host_vars variables for all nodes, only those of the currently running node.

```
...
some_info,
node1 : { node facts },
node2 : { node facts },
...
nodeN : { node facts },
some_info,
group_vars : { group variables },
var1 : "value 1", # Value from .../host_vars/current_node.yml
...
```

The latter does indeed include all the node's host_vars variables as you mentioned earlier.

```
...
node facts,
'var1': 'value1' # As set in .../host_vars/some_host.yml  },
...
```

So to me, that is very confusing since `hostvars['some_host']` seems to be a part of `hostvars`, but that's not the case. I don't know if that's a bug or a feature though?

So, if one needs to access all nodes' host_vars variables, he must loop on the nodes list somehow (using the list of the hosts involved in the play `vars.play_hosts` might be a good option) and gather the desired value node by node.

Thanks for your help Brian.
